components/core/tests/test-ffi_utils.cpp

```diff
@@ -0,0 +1,122 @@
+            msgpack_bytes.size()
+    );
+
+    optional<string> ret_val;
+    auto const& msgpack_obj{msgpack_oh.get()};
+    if (msgpack::type::ARRAY == msgpack_obj.type) {
+        if (false == serialize_and_append_msgpack_array_to_json_str(msgpack_obj, ret_val.emplace()))
+        {
+            return std::nullopt;
+        }
+    } else if (msgpack::type::MAP == msgpack_obj.type) {
+        if (false == serialize_and_append_msgpack_map_to_json_str(msgpack_obj, ret_val.emplace())) {
+            return std::nullopt;
+        }
+    } else {
+        return std::nullopt;
+    }
+    return ret_val;
+}
+}  // namespace
+
+TEST_CASE("test_msgpack_to_json", "[ffi][utils]") {
+    optional<string> result;
+    constexpr string_view cStringWithEscape{"String with\\\"escaped\"\\ characters\n\t"};
+
+    // Test array with primitive values only
+    json const array_with_primitive_values_only
+            = {1,
+               -1,
+               1.01,
+               -1.01,
+               true,
+               false,
+               "short_string",
+               "This is a long string",
+               cStringWithEscape,
+               nullptr};
+    result = serialize_msgpack_bytes_to_json_str(json::to_msgpack(array_with_primitive_values_only)
+    );
+    REQUIRE((result.has_value() && array_with_primitive_values_only == json::parse(result.value()))
+    );
+
+    // Test map with primitive values only
+    json const map_with_primitive_values_only
+            = {{"int_key", 1},
+               {"int_key_negative", -1},
+               {"float_key", 0.01},
+               {"float_key_negative", -0.01},
+               {"bool_key_true", false},
+               {"bool_key_false", true},
+               {"str_key", "Test string"},
+               {"str_with_\"escaped\"_key", cStringWithEscape},
+               {"null_key", nullptr}};
+    result = serialize_msgpack_bytes_to_json_str(json::to_msgpack(map_with_primitive_values_only));
+    REQUIRE((result.has_value() && map_with_primitive_values_only == json::parse(result.value())));
+
+    // Test array with inner map
+    json array_with_map = array_with_primitive_values_only;
+    array_with_map.emplace_back(map_with_primitive_values_only);
+    result = serialize_msgpack_bytes_to_json_str(json::to_msgpack(array_with_map));
+    REQUIRE((result.has_value() && array_with_map == json::parse(result.value())));
+
+    // Test map with inner array
+    json map_with_array = map_with_primitive_values_only;
+    map_with_array.emplace("array_key", array_with_primitive_values_only);
+    result = serialize_msgpack_bytes_to_json_str(json::to_msgpack(map_with_array));
+    REQUIRE((result.has_value() && map_with_array == json::parse(result.value())));
+
+    // Recursively create inner maps and arrays
+    // Note: the execution time and memory consumption will grow exponentially as we increase the
+    // recursive depth.
+    constexpr size_t cRecursiveDepth{6};
```

Comment:
This depth is chosen to not take too long. `Depth == 6` takes 3 seconds on my laptop, whereas `Depth == 7` takes 25 seconds (which might be too long).